libfreerdp/codec/h264_mediacodec.c

```diff
@@ -0,0 +1,516 @@
+
+static int load_libmediandk(H264_CONTEXT* h264)
+{
+    H264_CONTEXT_MEDIACODEC* sys = (H264_CONTEXT_MEDIACODEC*)h264->pSystemData;
+    WLog_Print(h264->log, WLOG_INFO, "MediaCodec Loading libmediandk.so");
+
+    sys->mediandkLibrary = LoadLibraryA("libmediandk.so");
+    if (sys->mediandkLibrary == NULL) 
+    {
+        WLog_Print(h264->log, WLOG_ERROR, "Error loading libmediandk.so");
+        return -1;
+    }
+
+#define RESOLVE_MEDIANDK_FUNC(name) \
+    sys->fn##name = GetProcAddress(sys->mediandkLibrary, #name); \
+    if (sys->fn##name == NULL) \
+    { \
+        WLog_Print(h264->log, WLOG_ERROR, "Error resolving function " #name " from libmediandk.so"); \
+        return -1; \
+    }
+
+    const char **temp = NULL;
+#define RESOLVE_MEDIANDK_VARIABLE(member, exported) \
+    temp = GetProcAddress(sys->mediandkLibrary, exported); \
+    if (temp == NULL) \
+    { \
+        WLog_Print(h264->log, WLOG_ERROR, "Error resolving variable " exported " from libmediandk.so"); \
+        return -1; \
+    } \
+    sys->member = *temp;
+
+    RESOLVE_MEDIANDK_FUNC(AMediaFormat_new);
+    RESOLVE_MEDIANDK_FUNC(AMediaFormat_delete);
+    RESOLVE_MEDIANDK_FUNC(AMediaFormat_toString);
+    RESOLVE_MEDIANDK_FUNC(AMediaFormat_setInt32);
+    RESOLVE_MEDIANDK_FUNC(AMediaFormat_setString);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_createDecoderByType);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_delete);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_configure);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_start);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_stop);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_getInputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_getOutputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_dequeueInputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_queueInputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_dequeueOutputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_getOutputFormat);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_releaseOutputBuffer);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_setParameters);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_getName);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_releaseName);
+    RESOLVE_MEDIANDK_FUNC(AMediaCodec_getInputFormat);
+
+    RESOLVE_MEDIANDK_VARIABLE(gAMediaFormatKeyMime, "AMEDIAFORMAT_KEY_MIME");
+    RESOLVE_MEDIANDK_VARIABLE(gAMediaFormatKeyWidth, "AMEDIAFORMAT_KEY_WIDTH");
+    RESOLVE_MEDIANDK_VARIABLE(gAMediaFormatKeyHeight, "AMEDIAFORMAT_KEY_HEIGHT");
+    RESOLVE_MEDIANDK_VARIABLE(gAMediaFormatKeyColorFormat, "AMEDIAFORMAT_KEY_COLOR_FORMAT");
+
+    return 0;
+}
+
+static void unload_libmediandk(H264_CONTEXT* h264)
+{
+    H264_CONTEXT_MEDIACODEC* sys = (H264_CONTEXT_MEDIACODEC*)h264->pSystemData;
+    if (NULL == sys->mediandkLibrary)
+    {
+        return;
+    }
+
+    FreeLibrary(sys->mediandkLibrary);
+}
+
+static void set_mediacodec_format(H264_CONTEXT* h264, AMediaFormat** formatVariable, AMediaFormat* newFormat)
+{
+    H264_CONTEXT_MEDIACODEC* sys = (H264_CONTEXT_MEDIACODEC*)h264->pSystemData;
```

Comment:
Please `WINPR_ASSERT(h264)` and others before accessing them without `NULL` check.